web-server/v0.4/src/utils/request.js

```diff
@@ -20,79 +19,32 @@ const codeMessage = {
   503: 'The service is unavailable and the server is temporarily overloaded or maintained. ',
   504: 'The gateway timed out. ',
 };
-function checkStatus(response) {
-  if (response.status >= 200 && response.status < 300) {
-    return response;
-  }
+
+const errorHandler = error => {
+  const { response = {} } = error;
   const errortext = codeMessage[response.status] || response.statusText;
+  const { status, url } = response;
+
   notification.error({
-    message: `Request error ${response.status}: ${response.url}`,
+    message: `Request Error ${status}: ${url}`,
     description: errortext,
   });
-  const error = new Error(errortext);
-  error.name = response.status;
-  error.response = response;
-  throw error;
-}
 
-/**
- * Requests a URL, returning a promise.
- *
- * @param  {string} url       The URL we want to request
- * @param  {object} [options] The options we want to pass to "fetch"
- * @return {object}           An object containing either "data" or "err"
- */
-export default function request(url, options) {
-  const defaultOptions = {};
-  const newOptions = { ...defaultOptions, ...options };
-  if (
-    newOptions.method === 'POST' ||
-    newOptions.method === 'PUT' ||
-    newOptions.method === 'DELETE'
-  ) {
-    if (!(newOptions.body instanceof FormData)) {
-      newOptions.headers = {
-        Accept: 'application/json',
-        'Content-Type': 'application/json; charset=utf-8',
-        ...newOptions.headers,
-      };
-      newOptions.body = JSON.stringify(newOptions.body);
-    } else {
-      // newOptions.body is FormData
-      newOptions.headers = {
-        Accept: 'application/json',
-        ...newOptions.headers,
-      };
-    }
+  if (status === 403) {
+    router.push('/exception/403');
+    return;
+  }
+  if (status <= 504 && status >= 500) {
+    router.push('/exception/500');
+    return;
   }
+  if (status >= 404 && status < 422) {
+    router.push('/exception/404');
```

Comment:
So if we get a 404 for a `result.json` file, will this cause the user to go to another page?

Reply:
Correct; however, this doesn't seem to be the most ideal UX in this case. Ideally, we should redirect the user back to their previous "state" in the dashboard workflow. For example, if the user receives a 404 for a `result.json` file when navigating to the Summary page, they would be redirected back to the Results page. Since we receive this error within the component lifecycle of `componentDidMount()`, the user will see a loading indicator until the asynchronous network request has been completed and the request error handler is able to redirect the user back. 

I've included this solution in a separate commit and also removed `axios` as a dependency as `umi-request` is now the default networking library used throughout the dashboard.

Reply:
Have you thought about a simple pop-up window with the errors, with some context?

Reply:
<img width="1505" alt="Dashboard_-_Pbench_Dashboard" src="https://user-images.githubusercontent.com/4149652/57389362-fe26f380-7187-11e9-899c-e995aff73b66.png">

Yes! This is a pop-up window shown before redirecting the user back to their previous state within the dashboard session.